### ISSUE TITLE
Random errors or build hanging forever with task outputs derived from a configuration

### ISSUE THREAD

**Issue by emmanuelguerin** [gradle/gradle#17905]
Our company has a complex multi project build with custom plugins built in house. In some instances, the build seems
to hang after completing a task and waiting indefinitely, not running any more tasks.
We also have had some random failures:
```
java.util.ConcurrentModificationException: (No message provided)
at java.util.LinkedList$ListItr.checkForComodification(LinkedList.java:966)	
at java.util.LinkedList$ListItr.next(LinkedList.java:888)	
at org.gradle.execution.plan.DefaultExecutionPlan.selectNext(DefaultExecutionPlan.java:574)	
at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$executeNextNode$1(DefaultPlanExecutor.java:166)	
at org.gradle.internal.resources.DefaultResourceLockCoordinationService.withStateLock(DefaultResourceLockCoordinationService.java:45)	
at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:155)	
at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)	
at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)	
at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)	
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)	
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)	
at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)	
at java.lang.Thread.run(Thread.java:748)
```

### Expected Behavior
The build finishes (with error or not) and a message that explains what to change in the configuration if necessary.

### Current Behavior
Either failure on an unactionable error or build hanging forever.

When doing a thread dump on the builds hanging, I studied the Execution Worker and Daemon worker threads.
All execution workers are waiting on:
```
  java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at java.lang.Object.wait(Object.java:502)
	at org.gradle.internal.resources.DefaultResourceLockCoordinationService.withStateLock(DefaultResourceLockCoordinationService.java:51)
	- locked <0x00000005c00f5a30> (a java.lang.Object)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:155)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.lang.Thread.run(Thread.java:748)
```

The Daemon worker is stuck on:
```
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at java.lang.Object.wait(Object.java:502)
	at org.gradle.internal.resources.DefaultResourceLockCoordinationService.withStateLock(DefaultResourceLockCoordinationService.java:51)
	- locked <0x00000005c00f5a30> (a java.lang.Object)
	at org.gradle.execution.plan.DefaultPlanExecutor.awaitCompletion(DefaultPlanExecutor.java:83)
	at org.gradle.execution.plan.DefaultPlanExecutor.process(DefaultPlanExecutor.java:73)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph.executeWithServices(DefaultTaskExecutionGraph.java:193)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph.execute(DefaultTaskExecutionGraph.java:186)
	at org.gradle.execution.SelectedTaskExecutionAction.execute(SelectedTaskExecutionAction.java:39)
```

### Steps to Reproduce 
Unfortunately, I couldn't find a way to reproduce the problem outside of our main project. But it seems to only occur
when a particular set of tasks is in the task graph. Their particularity that no other tasks has is that they define a
method with @OutputFiles that returns a FileCollection derived from a configuration. Something that looks like:
```
class TaskWithProblem extends DefaultTask {
    @InputFiles
    ConfigurableFileCollection input = project.files()

    @OutputFiles
    def getOutputs() {
        return input.files.collect { project.file("build/${it.name}") }
    }

    @TaskAction
    void exec() {}
}

configurations.create("myconfig") {
    canBeResolved = true
    canBeConsumed = false
}

tasks.register("taskWithProblem", TaskWithProblem) {
    input.from(configurations.myconfig)
}
```
(Basically, that task creates one output file per input file, and we set its input to a configuration)
The problem happens even if the configuration is empty.

### Your Environment
It is difficult to say when the problem started to appear, because of its randomness. But it occurs for
sure in all Gradle 7* releases (tested up to 7.1.1), on a variety of environments (Windows, Mac and Linux).
Also happens with JDK 8 or 11

**Comment by emmanuelguerin** [comment]
I did some debugging and found the following strange behaviour:
- one execution worker reached the selectNext method of the execution plan and considers the taskWithProblem. The other workers seem to properly wait using the lock in DefaultResourceLockCoordinationService.withStateLock.
- that worker then attempts to resolveMutations on the node.
- This reaches the `DefaultConfiguration.resolveToStateOrLater` method where the `!owner.getModel().hasMutableState()` method returns false. It triggers a deprecation warning, but attempts the `owner.getModel().fromMutableState(p -> resolveExclusively(requestedState));`
- down the line, the thread reenters DefaultResourceLockCoordinationService.withStateLock that triggers a "wait()" on the lock, freeing all other threads to pursue their on selectNext methods.

**Comment by emmanuelguerin** [comment]
It seems this is the same as #17013

**Comment by mlopatkin** [comment]
The issue looks very similar indeed but I defer to the experts in this area. @gradle/execution could you please take a look at this?

**Comment by wolfs** [comment]
Yeah, I think this issue is a duplicate of #17013 and will be closed as such.

**Comment by emmanuelguerin** [comment]
Hello. We've just installed the 7.4 a week ago, switching from our own fork of 7.1. The deadlock seems gone. Thank you @wolfs !
